### DISCUSSION TITLE
<newbie> Server send just deltas?

### DISCUSSION THREAD

**Discussion by yatesco** [tbranyen/diffhtml#237] [Ideas]
Hi - if the server caches the clients current HTML and generates the new HTML, can the server use this library to determine just the deltas and have those sent to the client which then applies those deltas?

I didn't find anything obvious, but this seems an obvious and incredibly powerful (not implying cheap or expectations of who does the work ;-)) capability.

This might be obviously documented - apologies, I'm skim reading a bunch of these libraries ;-)

**Comment by tbranyen** [comment]
Middleware has the ability to halt, skip, and apply patches, so I think it could be doable from that approach. You would just need to be 100% sure the old markup matches the server, otherwise the deltas will be invalid.

**Reply by yatesco** [reply]
thanks Tim. So this capability is already surfaced in the middleware APIs? I'll take a look - thanks!

**Reply by tbranyen** [reply]
Yep, but it's not super obvious how to do it from the docs. Keep in mind this is uncharted territory, which is great, because diffHTML was designed to be customized and this pushes the bounds.

**Comment by tbranyen** [comment]
Essentially every render in diffHTML is a [Transaction](https://github.com/tbranyen/diffhtml/blob/master/packages/diffhtml/lib/transaction.js#L35) instance. These transactions run predictable tasks in order. Once the last task runs, the transaction/render is considered complete. You can see the tasks [here](https://github.com/tbranyen/diffhtml/blob/master/packages/diffhtml/lib/transaction.js#L32):

```js
export const tasks = {
  schedule, shouldUpdate, reconcileTrees, syncTrees, patchNode, endAsPromise,
};
```

For handling only deltas on the client, we want all the tasks, except for `syncTrees`, as the server side will be doing this part. So simply swapping out that call with a new task to inject the deltas should allow `patchNode` to work correctly.

The render APIs model the browser, so we have `outerHTML` and `innerHTML`, these take in a new rendered tree. Since we are using deltas, you should probably just pass `null`.

Here is a working solution of applying just delta/patches:

```js
import { Internals, innerHTML, toString, html, use } from 'diffhtml';
import { JSDOM } from 'jsdom';

// simulate browser
global.document = new JSDOM(`<!DOCTYPE html>`).window.document;

///////////////////////////////////////////////////////////////////////////////

const applyDeltasTask = config => transaction => {
  const currentTasks = transaction.tasks;
  const indexOfSyncTrees = currentTasks.indexOf(Internals.tasks.syncTrees);

  // Only run this middleware when deltas are present.
  if (!('deltas' in transaction.config)) {
    return;
  }

  // Replace syncTrees with applyDeltas
  currentTasks.splice(indexOfSyncTrees, 1, function applyDeltas(transaction) {
    transaction.patches = transaction.config.deltas.map(x => {
      if (x === Symbol.for('oldTree')) {
        return transaction.oldTree;
      }
      return x;
    });
  });
};

///////////////////////////////////////////////////////////////////////////////

const mount = document.createElement('div');

use(applyDeltasTask());

innerHTML(mount, null, {
  // Simulate deltas from the server, this just adds a hello world node
  deltas: [
    Internals.PATCH_TYPE.INSERT_BEFORE,
    Symbol.for('oldTree'),
    html`Hello world yeee!`,
    null,
  ],
});

console.log(toString(mount));

// <div>Hello world yeee!</div>
```

The most challenging part will be figuring out a way to align the references from the server to the client, specifically the `mount` variable below. If I were to tackle this, I would probably use some kind of `WeakMap` or similar to have a symbol passed in the deltas, and then remapped locally to the actual value. For now I've added a `Symbol.for('oldTree')` as an example, but this isn't robust.

```js
  deltas: [
    Internals.PATCH_TYPE.INSERT_BEFORE,
    mount, // Symbol.for('oldTree') or something more robust
    html`Hello world!`,
    null,
  ],
```

**Reply by tbranyen** [reply]
~~Hrm, for some reason this isn't working 100% yet. I thought it was, but I had a typo in the code. I'll come back to this later today or tomorrow. I think it's really close.~~ Fixed

**Comment by yatesco** [comment]
Many thanks! It’ll be next week before I get a chance to start on my prototype, but this is very encouraging, thanks.

**Reply by tbranyen** [reply]
Sweet! I'll continue to iterate on this concept too. If we can solve this properly, I think we'll get Web Worker support figured out as well. These are very much in the same problem-space.

**Comment by tbranyen** [comment]
I've been making some pretty amazing [progress for web worker support](https://github.com/tbranyen-experimental/diffhtml-web-worker/blob/main/index.js). One of the biggest revelations has been thinking about the client as simply the delta patcher and event handler (thank you for that!). This led to another revelation to migrate all of the actual application logic into the worker thread. Before I was trying to offload only the diffing, when in reality you need to offload the whole application. I've been able to churn out 60fps across multiple threads using this approach.

From a server-perspective, I can imagine replacing the postMessage web worker calls, with a web socket handler. There isn't much difference otherwise. They have the same serialization issues, are both bidirectional, and can support array buffers (for future optimizations).

I was thinking about the issue around event handlers, since the browser event system requires synchronous handling of event propagation. We have a module called [diffhtml-middleware-synthetic-events](https://github.com/tbranyen/diffhtml/tree/master/packages/diffhtml-middleware-synthetic-events). What if we allowed this to implement it's own event bubbling/propagation rules that allow asynchronous handling?

This would effectively make the client extremely thin. It would dispatch emulated DOM events and apply deltas.

**Comment by yatesco** [comment]
This sounds awesome! I’m still scaffolding my prototype but will hope to be ready to play late this week. Good work though - nice one.

**Comment by tbranyen** [comment]
I'm still making occasional progress on this original ask. I deviated a bit with web workers, but now I'm adapting that code into the server approach. One difference between the worker and the server is that the server needs to generate an entire HTML page, while the worker has the benefit of one already existing.

This is problematic when it comes to hydrating. With the worker approach, there is no hydration, the same code cranks out renders and applies deltas the same for the first to the last render. With the server you'll most likely want to send a fully rendered initial page, and then progressively send deltas. Since the initial render generates the `__link` ids required for aligning deltas, and we don't get that information in the rendered markup, it becomes very difficult to properly align.

We could do a few things to solve this:

- Create a method that can hash a VTree position into a deterministic id, like you'd see with hashing a file into a SHA, this is the approach React has taken with `useId`: https://github.com/reactwg/react-18/discussions/111
- Inline `__link` information directly into the markup, allowing the client to establish linkage and state directly from the page
- Not support an initial render, and only allow web socket deltas to be applied. This would work, but not be ideal at all from a performance/accessibility perspective

I have a feeling that `useId` will be the way we go forward since it is the way React is solving the problem. Since that is going to be complicated and require changes to the diffHTML internals, I'm going to start a proof-of-concept by inlining the references.

What I'd expect to see from all this work by the end of the year is a common middleware that can be used in web workers and nodejs servers to produce deltas and apply them in a client side main thread.

**Reply by yatesco** [reply]
Excellent. I feel guilty being a pure observer here, and unfortunately I've had a non-trivial health issue which has delayed things quite significantly, so it is unlikely I'm going to be able to do more than use the library as-is, if at all :-(

**Reply by tbranyen** [reply]
No sweat and I wish you a speedy recovery! No expectations any of this will be used beyond me. You just brought up an excellent topic and lit a spark by framing an implementation that I hadn't considered. It's been a good exercise and the progress is exciting.

**Reply by yatesco** [reply]
Bless you buddy.

**Comment by tbranyen** [comment]
Woot finally have the initial ask prototyped. Still missing events and DOM access, but for now, I have 60fps rendering a SVG clock using deltas via WebSockets. Check it out here: https://github.com/tbranyen-experimental/diffhtml-web-worker/blob/server/server.js. Full rendering occurs on the server. Memory is excellent! This could be pretty amazing for users with low-end devices. Kind of borrows from the technique used by companies streaming games, where they have an input (in our case dom/input events) system and streaming pre-rendered deltas.

**Reply by tbranyen** [reply]
![image](https://user-images.githubusercontent.com/181635/145700540-84425211-a67b-4564-80ae-81a6a071678d.png)

**Reply by tbranyen** [reply]
This is a screenshot from the client side looking at the only JS execution, which applies the patches.

**Reply by yatesco** [reply]
This is just fantastic buddy - well done, sincerely, and thank you :-)

**Comment by tbranyen** [comment]
Just posted another update that has some pretty interesting changes. One thing I realized is that no matter what, we're going to need synchronous access to properties on `window` and in the DOM. We can't do that on the main thread in Node, but if we could spin up a thread per WebSocket connection, and do the server diffing there, we could then proxy the patches from the server worker. This would make the `<${App} />` render flow look like:

**Server worker** -> **Server main thread** -> **Web Socket** --> **Main thread patches**

Since the server worker contains the calls to the `window` object and wants the value back synchronously, we can block there just like partytown does for web workers. We can probably use the exact same mechanism. What's sweet is that this can all use the same middleware. Worker on the client, worker on the server, server without a worker, all the use cases get one way of handling it.

Obviously it's still very rough, but once I get synchronous DOM handling, you could basically start porting apps to this architecture. Whichever fits your use case.

**Reply by yatesco** [reply]
nice :-)

**Comment by tbranyen** [comment]
I was just thinking about this could change the game for instant live reloads. Instead of dealing with server and client code reloading, you can do it all from the server. You would create an initial worker per client connection, _wait for code change from developer_, then create a new worker, and share the last root state from the old worker as the current root state for diffing in the new worker, then shutdown the old worker. The new worker will synchronize and determine the deltas from what is currently rendered. So if you change one line, you should only see a text patch applied over the socket.

**Reply by yatesco** [reply]
you are on fire! :-)

**Comment by tbranyen** [comment]
Turns out what we're working on is basically identical to [Phoenix Frameworks LiveView](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html) feature. They render HTML on the server and calculate diffs to send over web sockets. I'm going to spend some time researching their approach and finishing up the synchronous global lookups.

I've got some HMR progress as well, but going to learn some LiveView first and then pick this back up. I'm still experimenting with sharedarraybuffers and atomics to provide synchronous cross-thread access, but that is gonna be a big project unless we can somehow use partytown.

**Reply by yatesco** [reply]
That was the original inspiration for the request :-)

**Comment by yatesco** [comment]
Just a note to say I've evaluated a bunch of vdom libraries and this is the only one I can find that:
 - is nippity quick ;-)
 - retains scroll position
 - retains input focus
 - just plain works

I'm still sending the full page HTML from the server rather than fragments, but so far it is more than quick enough.

Thanks!
